[Dynamo] Fix staticmethods for FSDP

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100117

This PR fixes capturing static methods for FSDP-managed modules. Previously, if a static method was invoked using `self.<staticmethod>`, then Dynamo would pass `self` twice to the method, causing a graph break due to the method being "unsupported". This PR achieves this by checking for `staticmethod` and using `UserFunctionVariable` instead of `UserMethodVariable`, which handles the correct calling convention.

This fixes FSDP + PT2 on HuggingFace's `T5ForConditionalGeneration`, which otherwise reports an error like the following based on the most recent trunk:
```
Output 0 of AsStridedBackward0 is a view of a view which was created in no_grad mode and is being modified inplace with grad mode enabled.
```
This is in reference to the `scores` tensor in `scores += position_bias_masked` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L559)).

I am not clear if this PR's fix is actually masking a different problem though. I wonder if there are edge cases with respect to Dynamo resuming execution and input mutations. Possibly, this PR only side steps the problem because there is no more recompilation at the static method `_relative_position_bucket()` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L443)).

In `UserDefinedObjectVariable.var_getattr()`, there is an existing branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L395-L398
I am not clear on when this branch can be triggered since if `subobj` is a static method, it still takes the `FunctionTypes` branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L403-L404
To preserve backward compatibility, the current version of this PR only modifies this `FunctionTypes` branch to differentiate between `staticmethod` and not `staticmethod`.

The PR that added this `FunctionTypes` branch is https://github.com/pytorch/pytorch/pull/92050/, and I checked that the added test `test_torch_distributions_functions()` only exercises the non-`staticmethod` case (since `Independent.log_prob` is not a `staticmethod`).

The last commit in `pytorch` that touched the `staticmethod` branch before https://github.com/pytorch/pytorch/pull/92050/ was the move from the `torchdynamo` repo into `pytorch`, so I cannot easily tell which test cases it corresponds to.


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Commit:
[Dynamo] Fix staticmethods for FSDP

[ghstack-poisoned]

Commit:
Update on "[Dynamo] Fix staticmethods for FSDP"


This PR fixes capturing static methods for FSDP-managed modules. Previously, if a static method was invoked using `self.<staticmethod>`, then Dynamo would pass `self` twice to the method, causing a graph break due to the method being "unsupported". This PR achieves this by checking for `staticmethod` and using `GetAttrVariable` instead of `UserMethodVariable`, which handles the correct calling convention.

This fixes FSDP + PT2 on HuggingFace's `T5ForConditionalGeneration`, which otherwise reports an error like the following based on the most recent trunk:
```
Output 0 of AsStridedBackward0 is a view of a view which was created in no_grad mode and is being modified inplace with grad mode enabled.
```
This is in reference to the `scores` tensor in `scores += position_bias_masked` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L559)).

I am not clear if this PR's fix is actually masking a different problem though. I wonder if there are edge cases with respect to Dynamo resuming execution and input mutations. Possibly, this PR only side steps the problem because there is no more recompilation at the static method `_relative_position_bucket()` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L443)).

In `UserDefinedObjectVariable.var_getattr()`, there is an existing branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L395-L398
I am not clear on when this branch can be triggered since if `subobj` is a static method, it still takes the `FunctionTypes` branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L403-L404
To preserve backward compatibility, the current version of this PR only modifies this `FunctionTypes` branch to differentiate between `staticmethod` and not `staticmethod`.

The PR that added this `FunctionTypes` branch is https://github.com/pytorch/pytorch/pull/92050/, and I checked that the added test `test_torch_distributions_functions()` only exercises the non-`staticmethod` case (since `Independent.log_prob` is not a `staticmethod`).

The last commit in `pytorch` that touched the `staticmethod` branch before https://github.com/pytorch/pytorch/pull/92050/ was the move from the `torchdynamo` repo into `pytorch`, so I cannot easily tell which test cases it corresponds to.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

awgu commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -401,6 +401,15 @@ def var_getattr(self, tx, name):
         elif isinstance(subobj, types.FunctionType):
+            # Check `__dict__` to bypass the function descriptor protocol to
+            # accurately check for static method
+            is_staticmethod = name in type(self.value).__dict__ and isinstance(
```
If people know of a better way to check this, feel free to let me know.

wconstab replied:
maybe try something like this?
https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/nn_module.py#L222

awgu replied:
This does not seem to be correct :/

```
>>> import torch
>>> from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
>>> from torch._dynamo.utils import istype
>>> istype(FSDP.summon_full_params, staticmethod)
False
```
https://github.com/pytorch/pytorch/blob/5f138a6b6561cd6fa4653d06623bcf0a72b72fd0/torch/distributed/fsdp/fully_sharded_data_parallel.py#L792-L794

I checked in the branch itself with the new unit test too, and I see `istype(subobj, staticmethod)` returning `False`.

wconstab replied:
is it failing due to being wrapped in a contextmanager, or something else?

you may be able to use some utils from `inspect` directly to identify it's a staticmethod.

But my suggestion was also for the Variable part: did you consider using a UserFunctionVariable instead of a GetAttrVariable, as was done for those other staticmethods?

awgu replied:
> you may be able to use some utils from inspect directly to identify it's a staticmethod.

Unfortunately, I did not see anything relevant in `inspect`.

---

> is it failing due to being wrapped in a contextmanager, or something else?

I think it is something else since for the unit test the method is only a `staticmethod` without being a `contextmanager`.

I think what might be happening is that for `nn.Module`s, we do not use `inspect.getattr_static()`, so we do not bypass the descriptor protocol:
https://github.com/pytorch/pytorch/blob/88c8c2b71b7b2dac0b346dfadc15e96eacaa7755/torch/_dynamo/variables/user_defined.py#L347-L356
As a result, the `isinstance(subobj, staticmethod)` returns `False` unexpectedly.

---

> But my suggestion was also for the Variable part: did you consider using a UserFunctionVariable instead of a GetAttrVariable, as was done for those other staticmethods?

Let me look into this. I used `GetAttrVariable` because FSDP on T5 used to work before https://github.com/pytorch/pytorch/pull/92050, at which point I think the static method would fall through to the final `GetAttrVariable` case:
https://github.com/pytorch/pytorch/blob/88c8c2b71b7b2dac0b346dfadc15e96eacaa7755/torch/_dynamo/variables/user_defined.py#L443
(I was restoring the existing behavior.)

awgu replied:
For this PR's newly added branch, we pass `subobj` directly as the first arg to `UserFunctionVariable`, while for the existing `staticmethod` branch, we pass `subobj.__get__(self.value)`.

If we pass `subobj.__get__(self.value)` in this newly added branch, we hit something like
```
expected FunctionType found method <bound method TestSingleProc.test_fsdp_staticmethod.<locals>.ModuleWithStaticMethod._add of ModuleWithStaticMethod()>
```
where the first arg's type is incorrect.

Given this, it seems that the newly added branch and the existing `staticmethod` branch have different conditions and different args to `UserFunctionVariable`. I am not sure if we can coalesce them easily.

anijain2305 replied:
You are right about the `getattr_static` part. If we were not speciailzing for `nn.Modules`, we would have bypassed the descriptor protocol.

Commit:
Update on "[Dynamo] Fix staticmethods for FSDP"


This PR fixes capturing static methods for FSDP-managed modules. Previously, if a static method was invoked using `self.<staticmethod>`, then Dynamo would pass `self` twice to the method, causing a graph break due to the method being "unsupported". This PR achieves this by checking for `staticmethod` and using `GetAttrVariable` instead of `UserMethodVariable`, which handles the correct calling convention.

This fixes FSDP + PT2 on HuggingFace's `T5ForConditionalGeneration`, which otherwise reports an error like the following based on the most recent trunk:
```
Output 0 of AsStridedBackward0 is a view of a view which was created in no_grad mode and is being modified inplace with grad mode enabled.
```
This is in reference to the `scores` tensor in `scores += position_bias_masked` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L559)).

I am not clear if this PR's fix is actually masking a different problem though. I wonder if there are edge cases with respect to Dynamo resuming execution and input mutations. Possibly, this PR only side steps the problem because there is no more recompilation at the static method `_relative_position_bucket()` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L443)).

In `UserDefinedObjectVariable.var_getattr()`, there is an existing branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L395-L398
I am not clear on when this branch can be triggered since if `subobj` is a static method, it still takes the `FunctionTypes` branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L403-L404
To preserve backward compatibility, the current version of this PR only modifies this `FunctionTypes` branch to differentiate between `staticmethod` and not `staticmethod`.

The PR that added this `FunctionTypes` branch is https://github.com/pytorch/pytorch/pull/92050/, and I checked that the added test `test_torch_distributions_functions()` only exercises the non-`staticmethod` case (since `Independent.log_prob` is not a `staticmethod`).

The last commit in `pytorch` that touched the `staticmethod` branch before https://github.com/pytorch/pytorch/pull/92050/ was the move from the `torchdynamo` repo into `pytorch`, so I cannot easily tell which test cases it corresponds to.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

Commit:
Update on "[Dynamo] Fix staticmethods for FSDP"


This PR fixes capturing static methods for FSDP-managed modules. Previously, if a static method was invoked using `self.<staticmethod>`, then Dynamo would pass `self` twice to the method, causing a graph break due to the method being "unsupported". This PR achieves this by checking for `staticmethod` and using `UserFunctionVariable` instead of `UserMethodVariable`, which handles the correct calling convention.

This fixes FSDP + PT2 on HuggingFace's `T5ForConditionalGeneration`, which otherwise reports an error like the following based on the most recent trunk:
```
Output 0 of AsStridedBackward0 is a view of a view which was created in no_grad mode and is being modified inplace with grad mode enabled.
```
This is in reference to the `scores` tensor in `scores += position_bias_masked` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L559)).

I am not clear if this PR's fix is actually masking a different problem though. I wonder if there are edge cases with respect to Dynamo resuming execution and input mutations. Possibly, this PR only side steps the problem because there is no more recompilation at the static method `_relative_position_bucket()` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L443)).

In `UserDefinedObjectVariable.var_getattr()`, there is an existing branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L395-L398
I am not clear on when this branch can be triggered since if `subobj` is a static method, it still takes the `FunctionTypes` branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L403-L404
To preserve backward compatibility, the current version of this PR only modifies this `FunctionTypes` branch to differentiate between `staticmethod` and not `staticmethod`.

The PR that added this `FunctionTypes` branch is https://github.com/pytorch/pytorch/pull/92050/, and I checked that the added test `test_torch_distributions_functions()` only exercises the non-`staticmethod` case (since `Independent.log_prob` is not a `staticmethod`).

The last commit in `pytorch` that touched the `staticmethod` branch before https://github.com/pytorch/pytorch/pull/92050/ was the move from the `torchdynamo` repo into `pytorch`, so I cannot easily tell which test cases it corresponds to.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

Commit:
Update on "[Dynamo] Fix staticmethods for FSDP"


This PR fixes capturing static methods for FSDP-managed modules. Previously, if a static method was invoked using `self.<staticmethod>`, then Dynamo would pass `self` twice to the method, causing a graph break due to the method being "unsupported". This PR achieves this by checking for `staticmethod` and using `UserFunctionVariable` instead of `UserMethodVariable`, which handles the correct calling convention.

This fixes FSDP + PT2 on HuggingFace's `T5ForConditionalGeneration`, which otherwise reports an error like the following based on the most recent trunk:
```
Output 0 of AsStridedBackward0 is a view of a view which was created in no_grad mode and is being modified inplace with grad mode enabled.
```
This is in reference to the `scores` tensor in `scores += position_bias_masked` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L559)).

I am not clear if this PR's fix is actually masking a different problem though. I wonder if there are edge cases with respect to Dynamo resuming execution and input mutations. Possibly, this PR only side steps the problem because there is no more recompilation at the static method `_relative_position_bucket()` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L443)).

In `UserDefinedObjectVariable.var_getattr()`, there is an existing branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L395-L398
I am not clear on when this branch can be triggered since if `subobj` is a static method, it still takes the `FunctionTypes` branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L403-L404
To preserve backward compatibility, the current version of this PR only modifies this `FunctionTypes` branch to differentiate between `staticmethod` and not `staticmethod`.

The PR that added this `FunctionTypes` branch is https://github.com/pytorch/pytorch/pull/92050/, and I checked that the added test `test_torch_distributions_functions()` only exercises the non-`staticmethod` case (since `Independent.log_prob` is not a `staticmethod`).

The last commit in `pytorch` that touched the `staticmethod` branch before https://github.com/pytorch/pytorch/pull/92050/ was the move from the `torchdynamo` repo into `pytorch`, so I cannot easily tell which test cases it corresponds to.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -401,6 +401,16 @@ def var_getattr(self, tx, name):
+            if is_staticmethod:
+                # Use `GetAttrVariable` to avoid doubly passing in `self` as an
+                # argument, which happens if using `UserMethodVariable`
+                return variables.GetAttrVariable(self, name, source=source, **options)
```
Thanks for pointing this out. @awgu already did this.

anijain2305 reviewed (approved):
LGTM.

awgu commented:
@pytorchbot rebase -s

Commit:
Update on "[Dynamo] Fix staticmethods for FSDP"


This PR fixes capturing static methods for FSDP-managed modules. Previously, if a static method was invoked using `self.<staticmethod>`, then Dynamo would pass `self` twice to the method, causing a graph break due to the method being "unsupported". This PR achieves this by checking for `staticmethod` and using `UserFunctionVariable` instead of `UserMethodVariable`, which handles the correct calling convention.

This fixes FSDP + PT2 on HuggingFace's `T5ForConditionalGeneration`, which otherwise reports an error like the following based on the most recent trunk:
```
Output 0 of AsStridedBackward0 is a view of a view which was created in no_grad mode and is being modified inplace with grad mode enabled.
```
This is in reference to the `scores` tensor in `scores += position_bias_masked` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L559)).

I am not clear if this PR's fix is actually masking a different problem though. I wonder if there are edge cases with respect to Dynamo resuming execution and input mutations. Possibly, this PR only side steps the problem because there is no more recompilation at the static method `_relative_position_bucket()` ([code](https://github.com/huggingface/transformers/blob/a0ae2310ec46a2c592950babc85cf02e325bf6a7/src/transformers/models/t5/modeling_t5.py#L443)).

In `UserDefinedObjectVariable.var_getattr()`, there is an existing branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L395-L398
I am not clear on when this branch can be triggered since if `subobj` is a static method, it still takes the `FunctionTypes` branch:
https://github.com/pytorch/pytorch/blob/e5291e633fca9639ace4fb4585e1bb0b9dca9e95/torch/_dynamo/variables/user_defined.py#L403-L404
To preserve backward compatibility, the current version of this PR only modifies this `FunctionTypes` branch to differentiate between `staticmethod` and not `staticmethod`.

The PR that added this `FunctionTypes` branch is https://github.com/pytorch/pytorch/pull/92050/, and I checked that the added test `test_torch_distributions_functions()` only exercises the non-`staticmethod` case (since `Independent.log_prob` is not a `staticmethod`).

The last commit in `pytorch` that touched the `staticmethod` branch before https://github.com/pytorch/pytorch/pull/92050/ was the move from the `torchdynamo` repo into `pytorch`, so I cannot easily tell which test cases it corresponds to.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
